// @terminal-packages/sdk/dist/shared/enums/json-rpc-methods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var JsonRpcMethods;
(function (JsonRpcMethods) {
    JsonRpcMethods["personal_sign"] = "personal_sign";
    JsonRpcMethods["eth_accounts"] = "eth_accounts";
    JsonRpcMethods["eth_chainId"] = "eth_chainId";
    JsonRpcMethods["eth_sendTransaction"] = "eth_sendTransaction";
    JsonRpcMethods["eth_getTransactionReceipt"] = "eth_getTransactionReceipt";
    JsonRpcMethods["eth_sign"] = "eth_sign";
    JsonRpcMethods["eth_signTransaction"] = "eth_signTransaction";
    JsonRpcMethods["eth_sendRawTransaction"] = "eth_sendRawTransaction";
    JsonRpcMethods["eth_call"] = "eth_call";
    JsonRpcMethods["eth_getBalance"] = "eth_getBalance";
    JsonRpcMethods["eth_blockNumber"] = "eth_blockNumber";
    JsonRpcMethods["eth_gasPrice"] = "eth_gasPrice";
    JsonRpcMethods["eth_getTransactionCount"] = "eth_getTransactionCount";
    JsonRpcMethods["eth_getCode"] = "eth_getCode";
    JsonRpcMethods["eth_syncing"] = "eth_syncing";
    JsonRpcMethods["eth_protocolVersion"] = "eth_protocolVersion";
    JsonRpcMethods["eth_estimateGas"] = "eth_estimateGas";
    JsonRpcMethods["eth_getBlockByHash"] = "eth_getBlockByHash";
    JsonRpcMethods["eth_getBlockByNumber"] = "eth_getBlockByNumber";
    JsonRpcMethods["eth_getTransactionByHash"] = "eth_getTransactionByHash";
    JsonRpcMethods["eth_getBlockTransactionCountByHash"] = "eth_getBlockTransactionCountByHash";
    JsonRpcMethods["eth_getStorageAt"] = "eth_getStorageAt";
    JsonRpcMethods["eth_getBlockTransactionCountByNumber"] = "eth_getBlockTransactionCountByNumber";
    JsonRpcMethods["eth_getTransactionByBlockHashAndIndex"] = "eth_getTransactionByBlockHashAndIndex";
    JsonRpcMethods["eth_getTransactionByBlockNumberAndIndex"] = "eth_getTransactionByBlockNumberAndIndex";
    JsonRpcMethods["eth_getUncleByBlockHashAndIndex"] = "eth_getUncleByBlockHashAndIndex";
    JsonRpcMethods["eth_getUncleByBlockNumberAndIndex"] = "eth_getUncleByBlockNumberAndIndex";
    JsonRpcMethods["eth_getUncleCountByBlockHash"] = "eth_getUncleCountByBlockHash";
    JsonRpcMethods["eth_getUncleCountByBlockNumber"] = "eth_getUncleCountByBlockNumber";
    JsonRpcMethods["eth_getLogs"] = "eth_getLogs";
    JsonRpcMethods["eth_newFilter"] = "eth_newFilter";
    JsonRpcMethods["eth_newBlockFilter"] = "eth_newBlockFilter";
    JsonRpcMethods["eth_getFilterChanges"] = "eth_getFilterChanges";
    JsonRpcMethods["eth_getFilterLogs"] = "eth_getFilterLogs";
    JsonRpcMethods["eth_uninstallFilter"] = "eth_uninstallFilter";
    JsonRpcMethods["eth_newPendingTransactionFilter"] = "eth_newPendingTransactionFilter";
    JsonRpcMethods["eth_signTypedData"] = "eth_signTypedData";
    JsonRpcMethods["net_version"] = "net_version";
})(JsonRpcMethods = exports.JsonRpcMethods || (exports.JsonRpcMethods = {}));
