// @terminal-packages/sdk/dist/web3/providers/enums/source-categories.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SourceCategory;
(function (SourceCategory) {
    SourceCategory["Terminal"] = "RPC Endpoint";
    SourceCategory["Truffle"] = "SDK";
    SourceCategory["Alchemy"] = "RPC Endpoint";
    SourceCategory["MetaMask"] = "Wallet";
    SourceCategory["Infura"] = "RPC Endpoint";
    SourceCategory["Pocket"] = "SDK";
    SourceCategory["Ledger"] = "Wallet";
    SourceCategory["Web3ProviderEngine"] = "SDK";
    SourceCategory["WalletLink"] = "Wallet";
    SourceCategory["TruffleHDWalletProvider"] = "Wallet";
    SourceCategory["Portis"] = "Wallet";
    SourceCategory["QuikNode"] = "RPC Endpoint";
    SourceCategory["Torus"] = "Wallet";
    SourceCategory["Bitski"] = "Wallet";
    SourceCategory["Fortmatic"] = "Wallet";
    SourceCategory["SKALE"] = "Layer2";
    SourceCategory["Matic"] = "Layer2";
    SourceCategory["WalletConnect"] = "Wallet";
})(SourceCategory = exports.SourceCategory || (exports.SourceCategory = {}));
