// @terminal-packages/sdk/dist/ipfs/enums/multihash-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MultihashTypes;
(function (MultihashTypes) {
    MultihashTypes["sha1"] = "sha1";
    MultihashTypes["sha2_256"] = "sha2-256";
    MultihashTypes["sha2_512"] = "sha2-512";
    MultihashTypes["sha3"] = "sha3";
    MultihashTypes["sha3_224"] = "sha3-224";
    MultihashTypes["sha3_256"] = "sha3-256";
    MultihashTypes["sha3_384"] = "sha3-384";
    MultihashTypes["sha3_512"] = "sha3-512";
    MultihashTypes["dbl_sha2_256"] = "dbl-sha2-256";
    // infura dont support this right now
    MultihashTypes["murmur3_128"] = "murmur3-128";
    // not implemented by infura
    MultihashTypes["keccak_224"] = "keccak-224";
    // not implemented by infura
    MultihashTypes["keccak_226"] = "keccak-226";
    // not implemented by infura
    MultihashTypes["keccak_384"] = "keccak-384";
    MultihashTypes["keccak_512"] = "keccak-512";
    MultihashTypes["shake_128"] = "shake-128";
    MultihashTypes["shake_256"] = "shake-256";
    // not implemented by infura
    MultihashTypes["x11"] = "x11";
    MultihashTypes["md5"] = "md5";
})(MultihashTypes = exports.MultihashTypes || (exports.MultihashTypes = {}));
